File: bardecode/image_io.py
import json

import cv2
import numpy as np
from PIL import Image

from .schemas import Report, ImageResult


def read_image(path: str) -> np.ndarray:
    img = cv2.imread(path, cv2.IMREAD_COLOR)
    if img is None:
        img = _read_via_pillow(path)
        if img is None:
            raise FileNotFoundError(f"Cannot read image: {path}")
    return img


def _read_via_pillow(path: str) -> np.ndarray | None:
    try:
        pil = Image.open(path).convert("RGB")
        arr = np.array(pil)[:, :, ::-1].copy()
        return arr
    except Exception:
        return None


def format_report_json(report: Report, quiet: bool = False) -> str:
    if quiet:
        lines = []
        for ir in report.results:
            for b in ir.barcodes:
                lines.append(b.text)
        return "\n".join(lines)
    return json.dumps(_report_to_dict(report), ensure_ascii=False, indent=2)


def _report_to_dict(report: Report) -> dict:
    return {
        "tool": report.tool,
        "version": report.version,
        "results": [_image_result_to_dict(ir) for ir in report.results],
    }


def _image_result_to_dict(ir: ImageResult) -> dict:
    return {
        "image": ir.image,
        "width": ir.width,
        "height": ir.height,
        "barcodes": [
            {
                "format": b.format,
                "text": b.text,
                "bbox": list(b.bbox),
                "detection_score": b.detection_score,
            }
            for b in ir.barcodes
        ],
        "undecoded": list(ir.undecoded),
        "error": ir.error,
    }

File: cpp/src/image_io.cpp
#include "image_io.hpp"

#include <nlohmann/json.hpp>

namespace bardecode {

using json = nlohmann::json;

static json image_result_to_json(const ImageResult &ir) {
    json out;
    out["image"] = ir.image;
    out["width"] = ir.width;
    out["height"] = ir.height;
    out["barcodes"] = json::array();
    for (const auto &b : ir.barcodes) {
        out["barcodes"].push_back({
            {"format", b.format},
            {"text", b.text},
            {"bbox", b.bbox},
            {"detection_score", b.detection_score},
        });
    }
    out["undecoded"] = json::array();
    for (const auto &u : ir.undecoded) {
        out["undecoded"].push_back({
            {"bbox", u.bbox},
            {"detection_score", u.detection_score},
        });
    }
    out["error"] = ir.error ? json(*ir.error) : json(nullptr);
    return out;
}

std::string format_report_json(const Report &report, bool quiet) {
    if (quiet) {
        std::string lines;
        for (const auto &ir : report.results) {
            for (const auto &b : ir.barcodes) {
                if (!lines.empty()) lines += "\n";
                lines += b.text;
            }
        }
        return lines;
    }
    json j;
    j["tool"] = report.tool;
    j["version"] = report.version;
    j["results"] = json::array();
    for (const auto &ir : report.results) {
        j["results"].push_back(image_result_to_json(ir));
    }
    return j.dump(2);
}

}  // namespace bardecode
